### PR TITLE
Preliminary support for the 'Add Parameter' feature.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide Please review.

File: src/EditorFeatures/CSharpTest/AddParameter/AddParameterTests.cs [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+{
+    void M()
+    {
+        new C(true, 1);
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddParameter)]
+        public async Task TestAddAfterMatchingConstructorParam()
```
Comment (rchande) [inline src/EditorFeatures/CSharpTest/AddParameter/AddParameterTests.cs]: Worth adding tests for when multiple extra arguments are added?

File: src/Features/CSharp/Portable/GenerateConstructor/GenerateConstructorCodeFixProvider.cs [hunk 2]
```diff
@@ -26,10 +41,7 @@ internal class GenerateConstructorCodeFixProvider : AbstractGenerateMemberCodeFi
         private const string CS1503 = nameof(CS1503); // CS1503: Argument 1: cannot convert from 'T1' to 'T2'
         private const string CS7036 = nameof(CS7036); // CS7036: There is no argument given that corresponds to the required formal parameter 'v' of 'C.C(int)'
 
-        public override ImmutableArray<string> FixableDiagnosticIds
```
Comment (Pilchie) [inline src/Features/CSharp/Portable/GenerateConstructor/GenerateConstructorCodeFixProvider.cs]: The constants are still here too?

File: src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+            TObjectCreationExpressionSyntax objectCreation,
+            TArgumentSyntax argumentOpt)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+
+            // Not supported if this is "new Foo { ... }" (as there are no parameters at all.
+            var typeNode = syntaxFacts.GetObjectCreationType(objectCreation);
```
Comment (Pilchie) [inline src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs]: This doesn't seem like the right comment - (or it should be on that method).  There should be a type node in the example given...?

File: src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+
+            if (!type.IsNonImplicitAndFromSource())
+            {
+                return;
+            }
+
+            var arguments = (SeparatedSyntaxList<TArgumentSyntax>)syntaxFacts.GetArgumentsOfObjectCreationExpression(objectCreation);
+
+            var comparer = syntaxFacts.IsCaseSensitive
+                ? StringComparer.Ordinal
```
Comment (Pilchie) [inline src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs]: Technically, this should probably allow unicode escapes in identifiers to match...

File: src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+
+                    if (argumentToAdd != null)
+                    {
+                        if (argumentOpt != null && argumentToAdd != argumentOpt)
+                        {
+                            // We were trying to fix a specific argument, but the argument we want
+                            // to fix is something different.  That means there was an error earlier
+                            // than this argument.  Which means we're looking at a non-viable 
+                            // constructor.  Skip this one.
+                            continue;
```
Comment (Pilchie) [inline src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs]: Is there a test that covers this?  I didn't see it?

File: src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+                        constructorsAndArgumentToAdd.Add(
+                            (constructor, argumentToAdd, arguments.IndexOf(argumentToAdd)));
+                    }
+                }
+            }
+
+            // Order by the furthest argument index to the nearest argument index.  The ones with
+            // larger argument indexes mean that we matched more earlier arguments (and thus are
+            // likely to be the correct match).
+            foreach (var tuple in constructorsAndArgumentToAdd.OrderByDescending(t => t.index))
```
Comment (Pilchie) [inline src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs]: Is there a test that demonstrates what this sort is trying to do?

File: src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+
+            var methodDocument = invocationDocument.Project.Solution.GetDocument(methodDeclaration.SyntaxTree);
+            var newDocument = methodDocument.WithSyntaxRoot(newRoot);
+
+            return newDocument;
+        }
+
+        private static SyntaxNode GetNewMethodDeclaration(IMethodSymbol method, TArgumentSyntax argument, SeparatedSyntaxList<TArgumentSyntax> argumentList, SyntaxGenerator generator, SyntaxNode declaration, ISemanticFactsService semanticFacts, string argumentName, SyntaxNode expression, SemanticModel semanticModel, ITypeSymbol parameterType)
+        {
+            if (!string.IsNullOrWhiteSpace(argumentName))
```
Comment (Pilchie) [inline src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs]: Maybe rename to "inferredArgumentName" or something.

File: src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+            var root = methodDeclaration.SyntaxTree.GetRoot(cancellationToken);
+            var newRoot = root.ReplaceNode(methodDeclaration, newMethodDeclaration);
+
+            var methodDocument = invocationDocument.Project.Solution.GetDocument(methodDeclaration.SyntaxTree);
+            var newDocument = methodDocument.WithSyntaxRoot(newRoot);
+
+            return newDocument;
+        }
+
+        private static SyntaxNode GetNewMethodDeclaration(IMethodSymbol method, TArgumentSyntax argument, SeparatedSyntaxList<TArgumentSyntax> argumentList, SyntaxGenerator generator, SyntaxNode declaration, ISemanticFactsService semanticFacts, string argumentName, SyntaxNode expression, SemanticModel semanticModel, ITypeSymbol parameterType)
```
Comment (Pilchie) [inline src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs]: Wrap long lines.

File: src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+        {
+            if (!string.IsNullOrWhiteSpace(argumentName))
+            {
+                var newParameterSymbol = CodeGenerationSymbolFactory.CreateParameterSymbol(
+                    null, RefKind.None,
```
Comment (Pilchie) [inline src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs]: Is there never a case where we want to take refness into account?

File: src/Features/CSharp/Portable/GenerateConstructor/GenerateConstructorCodeFixProvider.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/GenerateConstructor/GenerateConstructorCodeFixProvider.cs]: Whoops.  Fixed.

File: src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs]: Ooops.  Fixing comment.

File: src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs]: All the checks are done to the 'interpreted' name.  so all escapes are gone at that point.

**Comment by CyrusNajmabadi** [comment]
> But I'd like a second review, and someone to try out this experience and bug bash it a bit before we merge.

I would absolutely love if someone wanted to bang on this a bit.  Does anyone from @dotnet/roslyn-ide want to take a look?  Note: this currently ony supports "add parameter" for the "new Foo(x, y, z)".  Methods are going to be more challenging and will come later.

**Comment by Pilchie** [comment]
Oh - I forgot to mention.  It seems like for constructors there should at least be the option of generating/finding a field/autoprop and assigning it.  Or were you thinking of it being a two-step process where you could go there and generate the hookup?

**Comment by CyrusNajmabadi** [comment]
I'm thinking the hookup would be an additional step.

**Comment by CyrusNajmabadi** [comment]
@jmarolf Yup!  This will move to methods as well.  Those get a bit more complicated though, so i wanted this general infrastructure in place.

**Pull request merged**